### PR TITLE
Tensor Pointwise ops

### PR DESCRIPTION
**Pull request by davoclavo**
This branch adds a handful of tensor ops wrappers listed on #21 - 

The current goal is to show some progress to see if I am on the right track. I would love to hear if you have any comments.

Some of the interesting challenges/thoughts so far:
 - [Implementation] Mapping accepted input and output dtypes has been interesting due to input restrictions and promotions - I wonder if there is an existing catalog of input/output dtypes for all operations.. perhaps in some other statically typed wrapper for libtorch
 - [Implementation] Handling `Tensor | Scalar` input invariants for a specific input argument
 - [Implementation] Define a more granular type hierarchy to be used in input/output restrictions
 - [Test] Floating point precision - perhaps adding an assertion helper that validates that float values are equal up to some precision, rather than the full value could ease adding tests by just importing values from the documentation
 - [Test] Thinking how can we validate if the input/output dtype type encodings we are implementing are indeed correct and avoid runtime errors (which are very cryptic) - probably with some kind of property testing?

Todo:
- [x] Handle output tensor tuples from native torch
- [x] Implement missing ops marked with TODO
- [x] Implement missing tests marked with TODO
 
Extra notes:
 - I am using devenv (nix) files to be able to handle system dependencies (see https://devenv.sh) -- I could also move those files to another PR, or delete them entirely if you think they shouldn't be in the main repo.

### CODE REVIEW

**Comment by sbrunk** [comment]
Thanks @davoclavo, this is great! Let me try to go through your comments one by one.

>  [Implementation] Mapping accepted input and output dtypes has been interesting due to input restrictions and promotions - I wonder if there is an existing catalog of input/output dtypes for all operations.. perhaps in some other statically typed wrapper for libtorch

Yeah this is tricky. I couldn't find anything in PyTorch.
Perhaps https://github.com/hasktorch/hasktorch/ as they seem to track the type as well.

Another idea that comes to mind is to use property based tests over dtypes to catch ops whose types should be narrowed down.

The design decision to track dtypes at compile has quite a lot of implications. Most implementations I've seen just have a raw tensor type. Sometimes I think it would have been easier to start with a raw tensor type as well and then add typed tensors and ops from there, i.e. `typed` package, with conversions.

>  [Implementation] Handling `Tensor | Scalar` input invariants for a specific input argument

Not sure if I get this right. Do you have an example? So far I've used overloading in a few places but that doesn't work in conjunction with default arguments.
 
>  [Implementation] Define a more granular type hierarchy to be used in input/output restrictions

Yeah I've given it some thought but then decided that I need to implement more ops to gain a better understanding how it should look like. :)

>  [Test] Floating point precision - perhaps adding an assertion helper that validates that float values are equal up to some precision, rather than the full value could ease adding tests by just importing values from the documentation

We can use `torch.allclose` here, see this test for instance:
https://github.com/sbrunk/storch/blob/667508b66a7a0e4cc54db839515e01b9136581e4/core/src/test/scala/torch/TensorSuite.scala#L81-L85

>  [Test] Thinking how can we validate if the input/output dtype type encodings we are implementing are indeed correct and avoid runtime errors (which are very cryptic) - probably with some kind of property testing?

This is closely related to your first point above. I agree that we should use property based testing for this.

> I am using devenv (nix) files to be able to handle system dependencies (see https://devenv.sh) -- I could also move those files to another PR, or delete them entirely if you think they shouldn't be in the main repo.

I have no experience with devenv but it's always good to have some kind of reproducable dev environment. Let me give it a try.

**Comment by sbrunk** [comment]
So I did have a look at the tests and realized we can already do some of this right now:

Let's say we implement the following op:
```scala
def multinomial[D <: DType](
    input: Tensor[D],
    numSamples: Long,
    replacement: Boolean = false
): Tensor[Int64] = Tensor(
  torchNative.multinomial(input.native, numSamples, replacement, GeneratorOptional())
)
```

We can add a property based test like so
```scala
class RandomSamplingSuite extends ScalaCheckSuite {

  property("multinomial") {
    forAll(genDType) { dtype =>
      val weights = torch.Tensor(Seq(0, 10, 3, 0)).to(dtype=dtype)
      val t = multinomial(weights, 2)
      // assertEquals(...) TODO
    }
  }

}
```

Which fails with a full stack trace (including the C++ stacktrace from the native code):

```
Exception raised on property evaluation.
> ARG_0: int16
> Exception: java.lang.RuntimeException: multinomial only supports floating-point dtypes for input, got: Short
Exception raised from multinomial_out at /Users/runner/work/javacpp-presets/javacpp-presets/pytorch/cppbuild/macosx-x86_64/pytorch/aten/src/ATen/native/Distributions.cpp:584 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, 
...
```

We can even look that check up in the [original C++ source](https://github.com/pytorch/pytorch/blob/2e2a74670dc231078666f8fbc16b63766ebe480b/aten/src/ATen/native/Distributions.cpp#L574-L606).

Now we already know we can restrict the DType of this op to floating point.

But let's try the test again with only float dtypes:

```diff
-  forAll(genDType) { dtype =>
+  forAll(genDType.filter(_.isInstanceOf[FloatNN])) { dtype => 
```

And we're still failing for float16:
```
Exception raised on property evaluation.
> ARG_0: float16
> Exception: java.lang.RuntimeException: multinomial is not implemented for half on CPU
```

Now it's getting interesting. It looks like fp16 is not supported on CPU, but on CUDA. I guess we need to bite the bullet and define `def multinomial[D <: FloatNN]` even though we know it can fail at runtime for fp16.

**Comment by sbrunk** [comment]
> Agree, it may be a lot of work to add all the possible mappings, but I think its awesome to add more compile-time checks, I think that's what pulled me into storch in the first place. 😃 - And for complicated things I think we could initially set them as the general `Tensor[DType]` | `Tensor[D]`

Yes absolutely, I agree it's great to have these compile-time checks to add a level of safety and to derive types.

But there are some things to be aware of. Here's just what I recall at the moment, there's probably more
- Since tensors are mutable (another interesting axis, JAX has only immutable tensors), we can't just make them covariant I think, so currently a `Tensor[Float32]` is *not* a `Tensor[DType]`, we can use `Tensor[?]` sometimes though.
- In a PyTorch module (which are also mutable) you can convert the parameter types of all submodules recursively. Now if we have a `val module MyModule[Float32]` and call `myModule.to(dtype=torch.float16)`, we break type-safety.
- What about things like quantization to 8/4 bit and [mixed-precision training](https://pytorch.org/docs/stable/amp.html) which are becoming more and more important with LLMs. I think these are inherently dynamic and I'm not sure yet how they interact with typed tensors.

So I think it's worth exploring tensor encodings a bit more. Perhaps we *do* need a raw tensor type, perhaps we can make the current type flexible enough, perhaps we need to look into mutability. Please don't hesitate to share your thoughts to improve this if you have any ideas any time.

What's important to me is to strike a good balance between usability and ease of porting models from PyTorch on the one hand, and type inference and safety on the other.

> Do you have any preference over **only** overloading, or are we OK with hybrid approach using union types for simple invariants that could have default arguments?

I'm totally fine with union types a well.

> Great idea! I expanded your thought and added a first stab to a generator for specific types. Will work on adding this to test for new tensor ops.
> 
> ```scala
>   val allDTypes = List(
>     int8,
>     uint8,
>     int16,
>     int32,
>     int64,
>     float32,
>     ...
>   )
> 
>   inline def genTensor[D <: DType]: Gen[Tensor[D]] =
>     Gen.oneOf(allDTypes.filter(_.isInstanceOf[D])).map { dtype =>
>       ones(10, dtype = dtype.asInstanceOf[D])
>     }
> ```
> 
> It would be awesome to be able to infer the different input types and just automagically provide generators for all the input types, but I guess that will be tricker, we could figure out how to do that in the future.

👍 

> Sweet! It's my first time working with git hooks, so not sure if this is a common/good practice: Should I add a `git-hooks/` folder for future contributors?
> 
> So they can just do
> 
> ```
> chmod +x git-hooks/pre-push-linting && cp git-hooks/pre-push-linting .git/hooks/
> ```

Great idea. I realized that as a pre-push hook it should probably just run the linting checks i.e. `sbt 'headerCheckAll ; scalafmtCheckAll ; scalafmtSbtCheck'` so you don't have to wait for the CI to catch it.

> Interesting! so these type limitations are specific to each device. I guess in the future more types could be added in the Tensor class, such as `Device`, and `Shape` (dimension sizes), and other important metadata

Yes there is some prior work in that direction like https://maximekjaer.github.io/tf-dotty/

> Lastly, in the spirit of having something accomplishable soon, and to not fall into an infinite rabbit hole, I think I will limit the scope of this PR to only add Pointwise Ops + their Tests (and leave generators and other operations for a separate PR)

Yes, let's do one step at a time 😄

File: core/src/test/scala/torch/TensorSuite.scala [hunk 1]
```diff
@@ -107,6 +145,114 @@ class TensorSuite extends ScalaCheckSuite {
...
+    assertEquals(
+      res,
+      Tensor(Seq(0.9186402074749388, 0.8630508763335075, -0.0259751641859841, -0.13000071184459938))
+    )
+  }
+
+  test("sinh") {
+    val t = Tensor(Seq(0.5380, -0.8632, -0.1265, 0.9399))
+    val res = sinh(t)
+    assertEquals(
```
Comment (sbrunk) [inline core/src/test/scala/torch/TensorSuite.scala:184]: You can use `tensor.allclose` here and for the other floating point comparisons

**Comment by sbrunk** [comment]
Hey @davoclavo thanks for pushing this further! I'm currently traveling but I'll have a closer look as soon as possible.

**Comment by sbrunk** [comment]
Yeah the tensor apply with complex numbers is also what's causing the CI build to still fail (due to fatal warnings enabled there):
```the type test for Array[spire.math.Complex[Double]] cannot be checked at runtime because its type arguments can't be determined from U```

I currently don't have good idea how to solve this as `ClassTag` knows only the top-level class of a type, so it works if `U` is a regular number type but not if it's a generic type itself like `Complex[T]. In Scala 2 this`TypeTag` probably would have been the solution but it's gone in Scala 3.

The shapeless solution looks interesting and I'm OK using it (perhaps until we find something else). I'm also fine with going through `Seq` for scalars as a workaround.

That being said: While it's great to have support for complex numbers, my personal priority currently is to support ops needed for building common neural network architectures, which usually operate on simple floats.

So if you don't have a specific need (or personal interest, I don't intend to discourage you here at all!) for supporting all operations on complex numbers, my suggestion is to give them lower priority going forward.

What do you think?

**Comment by sbrunk** [comment]
Wow @davoclavo fantastic work! Thanks so much for this.

The property-based testing enhancements are really useful. I think we could add your description about how to use property based testing to help implementing typed ops correctly the contributor docs.

The devenv right now is mostly about installing sbt and its dependencies like a JVM right? I think we should highlight in the contributor docs that the devenv is helpful but optional (you can also install sbt through other means like coursier).

But that's something that can be added later as well so LGTM.

**Comment by sbrunk** [comment]
> Just for context, I have close to zero experience with DL / pytorch, so I sincerely appreciate your guidance on what to focus on. I should probably re-implement some model architectures next, so I get a sense of what is missing in storch. If you have any pointers on some mid-level tasks to tackle next, and in general advise/resources on how to approach DL from a practical point of view, I am very open to hear it and get some guidance.

If you want to see an example of an architecture implementation, have a look at the Storch [resnet implementation](https://github.com/sbrunk/storch/blob/main/vision/src/main/scala/torchvision/models/resnet.scala) that you can try through the [example training script](https://github.com/sbrunk/storch/blob/main/examples/src/main/scala/ImageClassifier.scala). It is basically a 1:1 port of the implementation in [torchvision](https://github.com/pytorch/vision/blob/v0.15.2/torchvision/models/resnet.py) to enable loading pre-trained weights. I've added the required ops/modules along the way while porting it.

If you're new to DL, I can recommend https://course.fast.ai/ (they use PyTorch under the hood as well). They also have a book.

[Neural Networks: Zero to Hero](https://www.youtube.com/playlist?list=PLAqhIrjkxbuWI23v9cThsA9GvCAUhRvKZ) by Andrej Karpathy is also very hands-on.
